test/test_functionalization.py

```diff
@@ -104,10 +118,27 @@ def f(x):
 def forward(self, a_1):
     empty = torch.ops.aten.empty.memory_format([4, 2], dtype = torch.float32, device = device(type='cpu'), pin_memory = False)
     fill_scalar = torch.ops.aten.fill.Scalar(empty, 1.0);  empty = None
-    view_copy_default = torch.ops.aten.view_copy.default(a_1, [4, 2]);  a_1 = None
+    view_copy_default = torch.ops.aten.view_copy.default(a_1, [4, 2])
     add_tensor = torch.ops.aten.add.Tensor(view_copy_default, fill_scalar);  view_copy_default = fill_scalar = None
     view_copy_default_1 = torch.ops.aten.view_copy.default(add_tensor, [4, 2])
-    mul_tensor = torch.ops.aten.mul.Tensor(view_copy_default_1, view_copy_default_1);  view_copy_default_1 = None
+    mul_tensor = torch.ops.aten.mul.Tensor(view_copy_default_1, view_copy_default_1)
+    copy__default = torch.ops.aten.copy_.default(a_1, view_copy_default_1);  a_1 = view_copy_default_1 = None
+    return add_tensor
+    """)
+
+        reinplaced_logs = self.get_logs(f, torch.ones(4, 2), reapply_views=True, run_reinplace=True)
+        self.assertExpectedInline(reinplaced_logs, """\
+
+
+
+def forward(self, a_1):
+    empty = torch.ops.aten.empty.memory_format([4, 2], dtype = torch.float32, device = device(type='cpu'), pin_memory = False)
+    fill_scalar = torch.ops.aten.fill_.Scalar(empty, 1.0)
+    view_default = torch.ops.aten.view.default(a_1, [4, 2])
+    add_tensor = torch.ops.aten.add.Tensor(view_default, empty);  view_default = empty = None
+    view_default_1 = torch.ops.aten.view.default(add_tensor, [4, 2])
+    mul_tensor = torch.ops.aten.mul.Tensor(view_default_1, view_default_1)
+    copy__default = torch.ops.aten.copy_.default(a_1, view_default_1);  a_1 = view_default_1 = None
     return add_tensor
```

Comment:
These expects are going to get a bit tiresome to audit by hand. Can we come up with some invariants to test? For example, I'd expect functionalize . reinplace (or vice versa) to be idempotent (maybe after an initial reinplace to deal with programs that aren't totally reinplaced as necessary). You'll need a way to test graph equivalence but IMO it's worth it.

Reply:
Another simple thing to test: do the graphs give the same outputs given the same inputs

Reply:
I'm pretty sure that... Once we (a) add the optimization for `copy_()`, and (b) add an extra pass / bit of logic to remove duplicate views, then the output of `reinplace` should be the same as the original like you're saying. Once those two changes are in, I'll test out removing these extra expect tests